src/nvim/eval/encode.c

```diff
@@ -579,21 +564,7 @@ static inline int convert_to_json_string(garray_T *const gap,
   } else {
     size_t utf_len = len;
     char *tofree = NULL;
-    if (last_p_enc != (const void *) p_enc) {
-      p_enc_conv.vc_type = CONV_NONE;
-      convert_setup(&p_enc_conv, p_enc, "utf-8");
-      p_enc_conv.vc_fail = true;
-      last_p_enc = p_enc;
-    }
-    if (p_enc_conv.vc_type != CONV_NONE) {
-      tofree = string_convert(&p_enc_conv, buf, &utf_len);
-      if (tofree == NULL) {
-        emsgf(_("E474: Failed to convert string \"%.*s\" to UTF-8"),
-              utf_len, utf_buf);
-        return FAIL;
-      }
-      utf_buf = tofree;
-    }
+    assert(STRCMP(p_enc, "utf-8") == 0);
```

Comment:
There is nothing that can change here, that encoding is UTF-8 is a _static_ property of the codebase. `p_enc` is mutable only because that is the simplest way to allow `set enc=utf-8` and synonyms but not `set enc=whatever`, a change of that value would indicate only a failure of option validation logic, not that the encoding suddenly is not UTF-8 anymore, which is impossible. We could make this clearer maybe by changing that to use `p_enc_dummy` and making `p_enc` a constant, not by adding this assert everywhere UTF-8 is assumed.